### PR TITLE
SONARJS-488 Migrate line metric

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-squid/src/main/java/org/sonar/javascript/metrics/FileLinesVisitor.java [hunk 1]
```diff
@@ -82,6 +81,7 @@ public void leaveFile(AstNode astNode) {
 
     if (fileLinesContext != null) {
 
+      // todo
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/metrics/FileLinesVisitor.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* CRITICAL 2 critical
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Remove this unused "FUNCTION_NODES" private field.](https://github.com/SonarCommunity/sonar-javascript/blob/057ab164b3b3873b3252ba3d8b86966c99b484a2/javascript-squid/src/main/java/org/sonar/javascript/JavaScriptAstScanner.java#L44) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)
* CRITICAL [Make this array "private".](https://github.com/SonarCommunity/sonar-javascript/blob/057ab164b3b3873b3252ba3d8b86966c99b484a2/javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java#L52) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1873)
* MAJOR [Rename "functionComplexity" which hides the field declared at line 62.](https://github.com/SonarCommunity/sonar-javascript/blob/057ab164b3b3873b3252ba3d8b86966c99b484a2/javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java#L129) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)
* CRITICAL [Make this array "private".](https://github.com/SonarCommunity/sonar-javascript/blob/057ab164b3b3873b3252ba3d8b86966c99b484a2/javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java#L43) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1873)

File: javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java [hunk 2]
```diff
@@ -76,6 +87,18 @@ public MetricsVisitor(FileSystem fs, SensorContext context) {
   }
 
   @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(CLASS_NODES)) {
+      classComplexity += getContext().getComplexity(tree);
+
+    } else if (tree.is(FUNCTION_NODES)) {
+      int functionComplexity = getContext().getComplexity(tree);
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java]: MAJOR Rename "functionComplexity" which hides the field declared at line 70. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      for (SyntaxTrivia trivia : ((SyntaxToken) tree).trivias()) {
+        if ((ignoreHeaderComments && seenFirstToken) || !ignoreHeaderComments) {
+          String[] commentLines = commentAnalyser.getContents(trivia.comment())
+              .split("(\r)?\n|\r", -1);
+          int line = trivia.startLine();
+          for (String commentLine : commentLines) {
+            if (commentLine.contains("NOSONAR")) {
+              noSonarLines.add(line);
+            }
+            if (!commentAnalyser.isBlank(commentLine)) {
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)

File: javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Kind.TOKEN)) {
+      for (SyntaxTrivia trivia : ((SyntaxToken) tree).trivias()) {
+        if ((ignoreHeaderComments && seenFirstToken) || !ignoreHeaderComments) {
+          String[] commentLines = commentAnalyser.getContents(trivia.comment())
+              .split("(\r)?\n|\r", -1);
+          int line = trivia.startLine();
+          for (String commentLine : commentLines) {
+            if (commentLine.contains("NOSONAR")) {
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* CRITICAL 2 critical
* MAJOR 4 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/javascript' package (org.sonarsource.javascript:javascript-squid:src/main/java/org/sonar/javascript) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/javascript/metrics' package (org.sonarsource.javascript:javascript-squid:src/main/java/org/sonar/javascript/metrics) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* CRITICAL [Make this array "private".](https://github.com/SonarCommunity/sonar-javascript/blob/34127a2ac2272f4b54d15e4db245015899e2fd96/javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java#L57) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1873)
* CRITICAL [Make this array "private".](https://github.com/SonarCommunity/sonar-javascript/blob/34127a2ac2272f4b54d15e4db245015899e2fd96/javascript-squid/src/main/java/org/sonar/javascript/metrics/MetricsVisitor.java#L48) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1873)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/plugins/javascript' package (org.sonarsource.javascript:sonar-javascript-plugin:src/main/java/org/sonar/plugins/javascript) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

File: javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+              noSonarLines.add(line);
+            }
+            if (!commentAnalyser.isBlank(commentLine)) {
+              comments.add(line);
+            }
+            line++;
+          }
+        } else {
+          seenFirstToken = true;
```
Comment (lindamartin) [inline javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java]: @vilchik-elena wha is the exact purpose of this variable ? I am not sure to understand ?

File: javascript-squid/src/test/java/org/sonar/javascript/JavaScriptAstScannerTest.java [hunk 6]
```diff
@@ -19,42 +19,16 @@
...
-    assertThat(file.getInt(EcmaScriptMetric.COMMENT_LINES)).isEqualTo(3);
-    assertThat(file.getNoSonarTagLines()).contains(10);
-    assertThat(file.getNoSonarTagLines().size()).isEqualTo(1);
-  }
-
-  @Test
-  public void lines() {
-    SourceFile file = JavaScriptAstScanner.scanSingleFile(new File("src/test/resources/metrics/lines.js"));
-    assertThat(file.getInt(EcmaScriptMetric.LINES)).isEqualTo(5);
+//    AstScanner<LexerlessGrammar> scanner = JavaScriptAstScanner.create(new EcmaScriptConfiguration(Charsets.UTF_8));
```
Comment (lindamartin) [inline javascript-squid/src/test/java/org/sonar/javascript/JavaScriptAstScannerTest.java]: @vilchik-elena is this test expected to be commented-out ?

File: javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java (hunk 5 above)
Comment (vilchik-elena) [inline javascript-squid/src/main/java/org/sonar/javascript/metrics/CommentLineVisitor.java]: @lindamartin **seenFirstToken** is required to track header comments (header comments are saved as trivias of first non-trivia token)

File: javascript-squid/src/test/java/org/sonar/javascript/JavaScriptAstScannerTest.java (hunk 6 above)
Comment (vilchik-elena) [inline javascript-squid/src/test/java/org/sonar/javascript/JavaScriptAstScannerTest.java]: @lindamartin i was not sure how to save this metric (FILES), and commented this out. But it seems that this metric is counted by platform side (as it-plugin runs well and has assert on this metric). So i will just drop this test

**Pull request closed without merging**